File: web/apps/suite/src/lib/mail/list-headers.ts
/**
 * Pure parsing helpers for the RFC 2369 `List-*` header family and the
 * RFC 8058 one-click unsubscribe extension.
 *
 * Requirements: docs/design/web/requirements/16-mailing-lists.md
 * (REQ-LIST-01/02, REQ-LIST-20..22) and
 * docs/design/web/requirements/14-unsubscribe.md (REQ-UNS-01..23).
 *
 * Nothing here touches the network or the DOM -- these are string ->
 * data-shape transforms so they are exhaustively unit-testable.
 */

/** One URL extracted from an angle-bracketed `List-*` header value. */
export interface ListUrl {
  scheme: 'https' | 'http' | 'mailto';
  url: string;
}

/**
 * Extract every `<...>` bracketed URL from a raw header value, tagging
 * each with its scheme. Values with an unrecognised scheme (rare, e.g.
 * a bare `NO` or a non-URL token) are silently dropped -- callers treat
 * an empty result the same as an absent header.
 *
 * RFC 2369 headers may carry several alternatives separated by commas;
 * the regex scan finds all of them regardless of the separator, so it
 * tolerates both comma-separated and space-separated forms seen in the
 * wild.
 */
export function parseAngleBracketUrls(raw: string | null | undefined): ListUrl[] {
  if (!raw) return [];
  const out: ListUrl[] = [];
  const re = /<([^>]+)>/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(raw))) {
    const url = (m[1] ?? '').trim();
    if (/^https:/i.test(url)) out.push({ scheme: 'https', url });
    else if (/^http:/i.test(url)) out.push({ scheme: 'http', url });
    else if (/^mailto:/i.test(url)) out.push({ scheme: 'mailto', url });
  }
  return out;
}

/** The result of picking a single actionable URL from a `List-*` header. */
export type ListAction =
  | { kind: 'https'; url: string }
  | { kind: 'mailto'; url: string }
  /**
   * Header present but the only URL(s) advertised are cleartext `http:`.
   * REQ-UNS-04 / REQ-LIST-20: never auto-open these; the caller must
   * surface the "unencrypted link" message instead.
   */
  | { kind: 'http-only'; url: string };

/**
 * Pick the preferred action from a parsed URL list: HTTPS first, then
 * (optionally) mailto:, then a cleartext http: URL flagged as
 * 'http-only' so the caller can refuse to auto-open it. Returns null
 * when the header carried no recognised URL at all.
 */
export function pickPreferredAction(
  urls: ListUrl[],
  opts: { allowMailto: boolean },
): ListAction | null {
  const https = urls.find((u) => u.scheme === 'https');
  if (https) return { kind: 'https', url: https.url };
  if (opts.allowMailto) {
    const mailto = urls.find((u) => u.scheme === 'mailto');
    if (mailto) return { kind: 'mailto', url: mailto.url };
  }
  const http = urls.find((u) => u.scheme === 'http');
  if (http) return { kind: 'http-only', url: http.url };
  return null;
}

/** REQ-LIST-02: the list's display label plus its raw identifier. */
export interface ListIdInfo {
  /** The raw content of the angle brackets, e.g. "projectx-discuss.example.com". */
  id: string;
  /** Human-readable label -- the description part, or a fallback. */
  label: string;
}

/**
 * Parse the `List-ID` header per REQ-LIST-02:
 * `"Project X discuss" <projectx-discuss.example.com>` -> label
 * "Project X discuss", id "projectx-discuss.example.com". When no
 * quoted description is present, the label falls back to the local
 * part of the identifier (the segment before the first dot).
 *
 * Returns null when the header is absent/blank or carries no
 * angle-bracketed identifier at all (a malformed header is treated the
 * same as an absent one -- REQ-LIST-01 says presence is the signal).
 */
export function parseListId(raw: string | null | undefined): ListIdInfo | null {
  if (!raw) return null;
  const trimmed = raw.trim();
  if (!trimmed) return null;
  const m = trimmed.match(/^(?:"((?:[^"\\]|\\.)*)"\s*)?<([^>]+)>/);
  if (!m) return null;
  const id = (m[2] ?? '').trim();
  if (!id) return null;
  const quoted = m[1];
  const label = quoted
    ? quoted.replace(/\\(.)/g, '$1').trim()
    : id.split('.')[0] || id;
  return { id, label: label || id };
}

/**
 * Parse `List-Post` per REQ-LIST-22. Returns the bare mailto address,
 * or null when the header is absent, blank, or the literal `NO`
 * reserved by RFC 2369 SS3.4 for "no posting allowed".
 */
export function parseListPostAddress(raw: string | null | undefined): string | null {
  const trimmed = (raw ?? '').trim();
  if (!trimmed) return null;
  if (/^NO$/i.test(trimmed)) return null;
  const bracketed = trimmed.match(/<mailto:([^>]+)>/i);
  if (bracketed) return bracketed[1] ?? null;
  const bare = trimmed.match(/^mailto:([^\s<>,]+)/i);
  if (bare) return bare[1] ?? null;
  return null;
}

/**
 * REQ-UNS-02: true when `List-Unsubscribe-Post` carries the RFC 8058
 * one-click marker. The header's only defined value is the literal
 * string `List-Unsubscribe=One-Click`; match case-insensitively since
 * some senders vary casing.
 */
export function hasOneClickPost(raw: string | null | undefined): boolean {
  if (!raw) return false;
  return /list-unsubscribe\s*=\s*one-click/i.test(raw);
}

/** The chosen unsubscribe mechanism per the `14-unsubscribe.md` priority order. */
export type UnsubscribeMechanism =
  | { kind: 'one-click'; url: string }
  | { kind: 'https'; url: string }
  | { kind: 'mailto'; url: string }
  | { kind: 'http-only'; url: string };

/**
 * Decide which unsubscribe mechanism to offer, given the raw
 * `List-Unsubscribe` and `List-Unsubscribe-Post` header values.
 *
 * Priority (REQ-UNS "Action: choose mechanism"):
 *   1. RFC 8058 one-click -- List-Unsubscribe-Post: One-Click plus an
 *      HTTPS List-Unsubscribe URL (REQ-UNS-20).
 *   2. Plain HTTPS URL (REQ-UNS-21).
 *   3. mailto: (REQ-UNS-22).
 *   4. A cleartext http: URL is never auto-actioned (REQ-UNS-04); it is
 *      returned as 'http-only' purely so the caller can still show the
 *      affordance (REQ-UNS-03 -- presence of *a* mechanism, even an
 *      unusable one, is what gates the button) and surface the warning
 *      on click instead of silently hiding.
 *
 * Returns null only when `List-Unsubscribe` is absent/carries no
 * recognised URL at all -- REQ-UNS-03's "no fallback" case.
 */
export function chooseUnsubscribeMechanism(
  listUnsubscribe: string | null | undefined,
  listUnsubscribePost: string | null | undefined,
): UnsubscribeMechanism | null {
  const urls = parseAngleBracketUrls(listUnsubscribe);
  if (urls.length === 0) return null;
  const oneClick = hasOneClickPost(listUnsubscribePost);
  const https = urls.find((u) => u.scheme === 'https');
  if (oneClick && https) return { kind: 'one-click', url: https.url };
  if (https) return { kind: 'https', url: https.url };
  const mailto = urls.find((u) => u.scheme === 'mailto');
  if (mailto) return { kind: 'mailto', url: mailto.url };
  const http = urls.find((u) => u.scheme === 'http');
  if (http) return { kind: 'http-only', url: http.url };
  return null;
}

/** The fields a `mailto:` URI can prefill in a compose window (REQ-UNS-22). */
export interface MailtoFields {
  to: string;
  subject: string;
  body: string;
}

/**
 * Parse a `mailto:` URI into compose fields per REQ-UNS-22 / REQ-LIST-21.
 * Tolerant of malformed percent-encoding -- falls back to the raw
 * (un-decoded) text rather than throwing.
 */
export function parseMailtoUri(uri: string): MailtoFields {
  const stripped = uri.replace(/^mailto:/i, '');
  const qIdx = stripped.indexOf('?');
  const addrPart = qIdx === -1 ? stripped : stripped.slice(0, qIdx);
  const queryPart = qIdx === -1 ? '' : stripped.slice(qIdx + 1);
  const to = safeDecode(addrPart);
  const params = new URLSearchParams(queryPart);
  return {
    to,
    subject: params.get('subject') ?? '',
    body: params.get('body') ?? '',
  };
}

function safeDecode(s: string): string {
  try {
    return decodeURIComponent(s);
  } catch {
    return s;
  }
}

File: mobile/shared/src/commonMain/kotlin/com/netzhansa/herold/shared/mail/ListHeaders.kt
package com.netzhansa.herold.shared.mail

/**
 * Parsing of the RFC 2369 `List-*` header family and the RFC 8058
 * one-click unsubscribe extension (suite `14-unsubscribe.md`
 * REQ-UNS-01..23). It mirrors the suite's `lib/mail/list-headers.ts` so
 * the two clients choose the same mechanism for the same message.
 *
 * Nothing here touches the network: these are string-to-shape transforms,
 * exercised in `commonTest`.
 */
object ListHeaders {
    /** One URL taken out of an angle-bracketed `List-*` header value. */
    data class ListUrl(val scheme: Scheme, val url: String)

    enum class Scheme { HTTPS, HTTP, MAILTO }

    /**
     * Every `<...>` bracketed URL of a raw header value, tagged with its
     * scheme. A value with an unrecognised scheme (a bare `NO`, a non-URL
     * token) is dropped, and a caller treats an empty result the same as
     * an absent header.
     */
    fun parseAngleBracketUrls(raw: String?): List<ListUrl> {
        if (raw.isNullOrBlank()) return emptyList()
        return BRACKETED.findAll(raw).mapNotNull { match ->
            val url = match.groupValues[1].trim()
            when {
                url.startsWith("https:", ignoreCase = true) -> ListUrl(Scheme.HTTPS, url)
                url.startsWith("http:", ignoreCase = true) -> ListUrl(Scheme.HTTP, url)
                url.startsWith("mailto:", ignoreCase = true) -> ListUrl(Scheme.MAILTO, url)
                else -> null
            }
        }.toList()
    }

    /**
     * REQ-UNS-02: true when `List-Unsubscribe-Post` carries the RFC 8058
     * one-click marker. Matched case-insensitively, since senders vary the
     * casing of the one defined value.
     */
    fun hasOneClickPost(raw: String?): Boolean {
        if (raw.isNullOrBlank()) return false
        return ONE_CLICK.containsMatchIn(raw)
    }

    /** The mechanism the Unsubscribe affordance runs. */
    sealed interface Mechanism {
        val url: String

        /** RFC 8058: an empty POST, no confirmation (REQ-UNS-20/30). */
        data class OneClick(override val url: String) : Mechanism

        /** A plain HTTPS URL, opened in the browser (REQ-UNS-21). */
        data class Https(override val url: String) : Mechanism

        /** A `mailto:`, which opens a prefilled compose (REQ-UNS-22). */
        data class Mailto(override val url: String) : Mechanism

        /**
         * Only a cleartext `http:` URL was advertised. The affordance is
         * still shown - the header names a mechanism (REQ-UNS-03) - and
         * acting on it surfaces the refusal rather than opening it
         * (REQ-UNS-04).
         */
        data class HttpOnly(override val url: String) : Mechanism
    }

    /**
     * The mechanism to offer for a message, given its two raw header
     * values. Priority: one-click, plain HTTPS, `mailto:`, cleartext.
     * Null when `List-Unsubscribe` names no recognised URL at all, which
     * is REQ-UNS-03's "no affordance, no body-text fallback".
     */
    fun chooseMechanism(listUnsubscribe: String?, listUnsubscribePost: String?): Mechanism? {
        val urls = parseAngleBracketUrls(listUnsubscribe)
        if (urls.isEmpty()) return null
        val https = urls.firstOrNull { it.scheme == Scheme.HTTPS }
        if (https != null) {
            return if (hasOneClickPost(listUnsubscribePost)) {
                Mechanism.OneClick(https.url)
            } else {
                Mechanism.Https(https.url)
            }
        }
        urls.firstOrNull { it.scheme == Scheme.MAILTO }?.let { return Mechanism.Mailto(it.url) }
        urls.firstOrNull { it.scheme == Scheme.HTTP }?.let { return Mechanism.HttpOnly(it.url) }
        return null
    }

    /** The fields a `mailto:` URI prefills in the composer (REQ-UNS-22). */
    data class MailtoFields(val to: String, val subject: String, val body: String)

    /**
     * Parses a `mailto:` URI into compose fields. Tolerant of malformed
     * percent-encoding: an undecodable segment is carried through as it
     * stands rather than failing the unsubscribe.
     */
    fun parseMailto(uri: String): MailtoFields {
        val stripped = uri.removePrefix("mailto:").removePrefix("MAILTO:")
        val queryIndex = stripped.indexOf('?')
        val addressPart = if (queryIndex == -1) stripped else stripped.substring(0, queryIndex)
        val queryPart = if (queryIndex == -1) "" else stripped.substring(queryIndex + 1)
        val params = queryPart.split('&').mapNotNull { pair ->
            if (pair.isBlank()) return@mapNotNull null
            val eq = pair.indexOf('=')
            if (eq == -1) return@mapNotNull null
            decode(pair.substring(0, eq)).lowercase() to decode(pair.substring(eq + 1))
        }.toMap()
        return MailtoFields(
            to = decode(addressPart),
            subject = params["subject"].orEmpty(),
            body = params["body"].orEmpty(),
        )
    }

    /**
     * Percent-decoding with `+` as a space, as a `mailto:` query carries
     * it. An incomplete escape is kept literally.
     */
    private fun decode(value: String): String {
        if (!value.contains('%') && !value.contains('+')) return value
        val bytes = mutableListOf<Byte>()
        var index = 0
        while (index < value.length) {
            val char = value[index]
            when {
                char == '+' -> {
                    bytes.add(' '.code.toByte())
                    index++
                }

                char == '%' && index + 2 < value.length + 1 -> {
                    val hex = value.drop(index + 1).take(2)
                    val parsed = hex.toIntOrNull(16)
                    if (hex.length == 2 && parsed != null) {
                        bytes.add(parsed.toByte())
                        index += 3
                    } else {
                        bytes.addAll(char.toString().encodeToByteArray().toList())
                        index++
                    }
                }

                else -> {
                    bytes.addAll(char.toString().encodeToByteArray().toList())
                    index++
                }
            }
        }
        return bytes.toByteArray().decodeToString()
    }

    private val BRACKETED = Regex("<([^>]+)>")
    private val ONE_CLICK = Regex("list-unsubscribe\\s*=\\s*one-click", RegexOption.IGNORE_CASE)
}
